tmt/steps/prepare/artifact/providers/koji.py

```diff
@@ -103,29 +110,67 @@ class KojiArtifactProvider(ArtifactProvider[RpmArtifactInfo]):
         artifacts = provider.download_artifacts(guest, Path("/tmp"), [])
     """
 
-    # TODO: Make RPM_BASE_URL configurable via FMF/CLI, not just env var
-    API_URL = os.getenv("KOJI_API_URL", "https://koji.fedoraproject.org/kojihub")  # For metadata
+    SUPPORTED_PREFIXES = ("koji.build:", "koji.task:", "koji.nvr:")
+
+    def __new__(cls, raw_provider_id: str, logger: tmt.log.Logger) -> 'KojiArtifactProvider':
+        """
+        Create a specific Koji provider based on the ``raw_provider_id`` prefix.
+
+        The supported provides are:
+        :py:class:`KojiBuild`,
+        :py:class:`KojiTask`,
+        :py:class:`KojiNvr`.
+
+        :raises ValueError: If the prefix is not supported
+        """
+        if raw_provider_id.startswith("koji.build:"):
+            return super().__new__(KojiBuild)
+        if raw_provider_id.startswith("koji.task:"):
+            return super().__new__(KojiTask)
+        if raw_provider_id.startswith("koji.nvr:"):
+            return super().__new__(KojiNvr)
+        # If we get here, the prefix is not supported
+        raise ValueError(
+            f"Unsupported artifact ID format: '{raw_provider_id}'. "
+            f"Supported formats are: 'koji.build:', 'koji.task:', 'koji.nvr:'"
+        )
 
     def __init__(self, raw_provider_id: str, logger: tmt.log.Logger):
         super().__init__(raw_provider_id, logger)
         self._session = self._initialize_session()
-        self._rpm_list = self._fetch_rpms()
+        self._build_provider: Optional[KojiBuild] = None
 
-    def _fetch_rpms(self) -> list[dict[str, Any]]:
+    @cached_property
```

Comment:
```suggestion
    @cached_property
    @abc.abstractmethod
```
Let's add the relevant abc hints.

Reply:
Refactored [9ed7722](https://github.com/teemtee/tmt/pull/4119/commits/9ed77227ce3d2485d23f08658fcd89449af13483)